### ISSUE TITLE
Path segments with percents encoding

### ISSUE THREAD

**Issue by jmini** [renatoathaydes/rawhttp#32]
Similar to #30, I have the feeling that the library is decoding path segments prior to sending the request.

---

URL encoding with [percents encoding](https://en.wikipedia.org/wiki/Percent-encoding) in path segments seems to be tricky.
For example it is not clear if `@` must be encoded or not in path-segments. E.g:`

```
http://localhost:1080/test/someone@example.com/test
http://localhost:1080/test/someone%40example.com/test
```

According to this Stackoverflow question: [URL encoding the character @ in query path](https://stackoverflow.com/questions/56498023/url-encoding-the-character-in-query-path) keeping the `@` not encoded is valid, but a lot of libs are replacing the `@` with`%40`.

It seems to me that RawHttp is doing the contrary: it is decoding the symbol prior to sending it.

```java
import java.net.Socket;

import rawhttp.core.RawHttp;
import rawhttp.core.RawHttpRequest;
import rawhttp.core.RawHttpResponse;

public class Snippet {
    public static void main(String[] args) throws Exception {
        String requestText = "GET /test/someone%40example.com/test HTTP/1.1\n" +
            "Host: localhost:1080\n" +
            "\n";
        RawHttp http = new RawHttp();
        RawHttpRequest httpRequest = http.parseRequest(requestText);
        try (Socket socket = new Socket("localhost", 1080)) {
            httpRequest.writeTo(socket.getOutputStream());
            RawHttpResponse<?> httpResponse = http.parseResponse(socket.getInputStream());
            String responseText = httpResponse.eagerly()
                .toString();
            socket.close();

            System.out.println(responseText);
        }
    }
}
```

Is arriving at the server side as `/test/someone@example.com/test`

----

Can be tested with [Mock-Server](https://www.mock-server.com/) by adding 2 expectations:

```
PUT http://localhost:1080/expectation
{
  "httpRequest" : {
    "method" : "GET",
    "path" : "/test/someone@example.com/test"
  },
  "httpResponse" : {
    "body" : "not_encoded"
  }
}

PUT http://localhost:1080/expectation
{
  "httpRequest" : {
    "method" : "GET",
    "path" : "/test/someone%40example.com/test"
  },
  "httpResponse" : {
    "body" : "is_encoded"
  }
}
```

When trying the Java Snippet, `is_encoded` is expected. Got `not_encoded`.

---

By comparison I have tried:

1) Firefox: working as expected, the client is not doing any de-encoding before sending the request.

2) [rest-client for vs-code](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) also working as expected. Those 2 requests are not the same:

```http
### Rest client init:
@base_url = http://localhost:1080

### Send request:
GET {{base_url}}/test/someone@example.com/test HTTP/1.1
```

```http
### Rest client init:
@base_url = http://localhost:1080

### Send request:
GET {{base_url}}/test/someone%40example.com/test HTTP/1.1
```

**Comment by jmini** [comment]
Here is my PR for this issue: https://github.com/renatoathaydes/rawhttp/pull/33

**Comment by renatoathaydes** [comment]
When you want to know the correct behaviour, you can always refer to the RFC which defines this.
In this case, the HTTP RFC refers to the URI RFC, which defines this:

https://tools.ietf.org/html/rfc3986#section-2.1

> A percent-encoding mechanism is used to represent a data octet in a
   component when that octet's corresponding character is outside the
   allowed set or is being used as a delimiter of, or within, the
   component.

In this case, the component is the "path": https://tools.ietf.org/html/rfc3986#section-3.3.

The problem is that paths are quite complicated... whether `@` is allowed depends on which kind of URI that is...

```
    path          = path-abempty    ; begins with "/" or is empty
                    / path-absolute   ; begins with "/" but not "//"
                    / path-noscheme   ; begins with a non-colon segment
                    / path-rootless   ; begins with a segment
                    / path-empty      ; zero characters

      path-abempty  = *( "/" segment )
      path-absolute = "/" [ segment-nz *( "/" segment ) ]
      path-noscheme = segment-nz-nc *( "/" segment )
      path-rootless = segment-nz *( "/" segment )
      path-empty    = 0<pchar>
      segment       = *pchar
      segment-nz    = 1*pchar
      segment-nz-nc = 1*( unreserved / pct-encoded / sub-delims / "@" )
                    ; non-zero-length segment without any colon ":"

      pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
```

Interestingly, the `@` is mentioned explicitly in 2 cases... but the only case we care about is the case where the URI contains a scheme (http<s>) and an authority... in which case we only need to look at these, I believe:

```
path-absolute = "/" [ segment-nz *( "/" segment ) ]
segment       = *pchar
segment-nz    = 1*pchar
pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
```

That means `@` is definitely allowed in a HTTP URL's path. However, we don't have to worry too much about this in RawHTTP because the Java standard library takes care of this, we just need to use it properly.

To confirm the above is true, I tried this Java code:

```java
import java.net.URI;

class Scratch {
    public static void main(String[] args) {
        System.out.println(URI.create("http://hello.com/foo/bar@boo").getRawPath());
        System.out.println(URI.create("http://hello.com/foo/bar@boo").getPath());
        // example with user access embedded in URL
        System.out.println(URI.create("http://m2@hello.com/foo/bar@boo").getRawPath());
    }
}
```

Prints:

```
/foo/bar@boo
/foo/bar@boo
/foo/bar@boo
```

But as I mentioned in your previous PR, I was going to get rid of that instantiation of `URI` as it became redundant... however, I want to optimise the library a little bit and stop using `toString()` to implement `writeTo` as I've seen that avoiding allocations in such writer methods can speed things up dramatically... so I have already modified the code you've changed. I will think a bit more about whether to accept the PR, but I think it's not necessary as the problem seems to not be going against the spec.

**Comment by jmini** [comment]
Thank you for the detailed explanations.

---

> I was going to get rid of that instantiation of `URI` as it became redundant

I think this is a good idea.

---

You can see at my issue the other way around:

If I have a `request.http` file like this:

```http
GET http://localhost:1080/test/someone%40example.com/test HTTP/1.1
```

It is not Ok for the RawHttp lib to send the request as:

`GET http://localhost:1080/test/someone@example.com/test`

Even if the URI RFC says the encoding of the `@` as `%40` is not necessary (from a URI semantic point of view).

At least other HTTP clients (such as Firefox (when you enter the URI in the address bar) or the the [rest-client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) inside Visual Studio Code) are not doing the optimization.

**Comment by renatoathaydes** [comment]
I agree, specially when RawHTTP's whole idea is that you know exactly what your HTTP messages will look like!

I just don't think this is a high priority issue... so I want to continue my refactoring of this part of the code to avoid wasting memory/allocations, but after I'm done I will make sure this issue has been fixed.

**Comment by renatoathaydes** [comment]
Released in 2.4.0.
